src/snowflake/snowpark/udf.py

```diff
@@ -897,3 +915,115 @@ def _do_register_udf(
                 )
 
         return UserDefinedFunction(func, return_type, input_types, udf_name)
+
+    def __do_register_udf(
+        self,
+        callableProperties: CallableProperties,
+        api_call_source: str,
+        stage_location: Optional[str] = None,
+        statement_params: Optional[Dict[str, str]] = None,
+        from_pandas_udf_function: bool = False,
+        skip_upload_on_content_match: bool = False,
+    ) -> UserDefinedFunction:
+
+        (
+            udf_name,
+            is_pandas_udf,
+            is_dataframe_input,
+            return_type,
+            input_types,
+        ) = process_registration_inputs(
+            session=self._session,
+            object_type=callableProperties.object_type,
+            func=callableProperties.func,
+            return_type=callableProperties.raw_return_type,
+            input_types=callableProperties.raw_input_types,
+            name=callableProperties.raw_name,
+        )
+
+        callableProperties.setValidatedObjectName(udf_name)
+        callableProperties.setValidatedReturnType(return_type)
+        callableProperties.setValidatedInputTypes(input_types)
+
+        arg_names = [f"arg{i + 1}" for i in range(len(input_types))]
+        callableProperties.setResolvedInputArgs(
+            [UDFColumn(dt, arg_name) for dt, arg_name in zip(input_types, arg_names)]
+        )
+
+        # allow registering pandas UDF from udf(),
+        # but not allow registering non-pandas UDF from pandas_udf()
+        if from_pandas_udf_function and not is_pandas_udf:
+            raise ValueError(
+                "You cannot create a non-vectorized UDF using pandas_udf(). "
+                "Use udf() instead."
+            )
+
```

Comment:
For lines 961 - 984, it breaks out the original `resolve_imports_and_packages` into self contained package and imports resolution.